src/vcpkg/base/system.mac.cpp

```diff
@@ -0,0 +1,144 @@
+#include <vcpkg/base/hash.h>
+#include <vcpkg/base/parse.h>
+#include <vcpkg/base/system.mac.h>
+#include <vcpkg/base/system.process.h>
+
+#include <vcpkg/metrics.h>
+
+#if defined(__linux__)
+#include <net/if.h>
+#include <netinet/in.h>
+#include <sys/ioctl.h>
+#endif
+
+namespace vcpkg
+{
+    Optional<StringView> find_first_nonzero_mac(StringView sv)
+    {
+        static constexpr StringLiteral ZERO_MAC = "00-00-00-00-00-00";
+
+        auto first = sv.begin();
+        const auto last = sv.end();
+
+        while (first != last)
+        {
+            // XX-XX-XX-XX-XX-XX
+            // 1  2  3  4  5  6
+            // size = 6 * 2 + 5 = 17
+            first = std::find_if(first, last, ParserBase::is_hex_digit);
+            if (last - first < 17)
+            {
+                break;
+            }
+
+            bool is_first = true;
+            bool is_valid = true;
+            auto end_of_mac = first;
+            for (int i = 0; is_valid && i < 6; ++i)
+            {
+                if (!is_first)
+                {
+                    if (*end_of_mac != '-')
+                    {
+                        is_valid = false;
+                        break;
+                    }
+                    ++end_of_mac;
+                }
+                is_first = false;
+
+                if (!ParserBase::is_hex_digit(*end_of_mac))
+                {
+                    is_valid = false;
+                    break;
+                }
+                ++end_of_mac;
+
+                if (!ParserBase::is_hex_digit(*end_of_mac))
+                {
+                    is_valid = false;
+                    break;
+                }
+                ++end_of_mac;
+            }
+            if (is_valid && StringView{first, end_of_mac} != ZERO_MAC)
+            {
+                return StringView{first, end_of_mac};
+            }
+            else
+            {
+                first = end_of_mac;
+            }
+        }
+
+        return nullopt;
+    }
+
+    std::string get_user_mac()
+    {
+        if (!LockGuardPtr<Metrics>(g_metrics)->metrics_enabled())
+        {
+            return "{}";
```

Comment:
Can the header have a comment as to what's expected for this stuff? I would argue that when metrics aren't enabled we shouldn't be calling a function named get_user_mac in the first place.

Reply:
Indicating that this returns a SHA rather than the actual thing would be nice too.

Reply:
agree, will make that change

Reply:
on the other hand this is called from `Metrics::enable()`...